### PR TITLE
Add allocateMemory helper with value known at compile time

### PR DESCRIPTION
**Pull request by axic**
As discussed in #6097.

### CODE REVIEW

**Comment by axic** [comment]
The remaining 4 places cannot be changed
```
                m_context << max(u256(32u), arrayType.memorySize());
                utils().allocateMemory();
```

because `memorySize()` is `u256`. I'm not sure why those have to be a bignum, when even if they exceed 32-bit would be infeasible gas wise.

We could also change `allocateMemory` to take `u256` now with the intention we clean up this later.

**Comment by axic** [comment]
Made the change, it takes `u256`. If happy, please squash before merge.

**Comment by axic** [comment]
The main point of this change is to avoid a `SWAP1` where possible, which should help the optimiser in some cases.

**Comment by axic** [comment]
Tests fail because there are gas savings.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1396,7 +1394,7 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		{
 			TypePointer arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
 			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
-			utils().allocateMemory(contract.name().length() + 32);
+			utils().allocateMemory(u256(contract.name().length() + 32));
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Why the explicit conversion?

**Review by Marenz — APPROVED** [review]

**Pull request merged**